src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,681 @@
+	MaxxStatus_t status;
+	uint32_t response_size = 0;
+
+	comp_info(dev, "waves_effect_message() start data %p size %d", data, size);
+
+	status = MaxxEffect_Message(waves_codec->effect, data, size,
+		waves_codec->response, &response_size);
+
+	if (status) {
+		comp_err(dev, "waves_effect_message() MaxxEffect_Message() error %d",
+			status);
+		return -EIO;
+	}
+
+	// at time of writing codec adapter does not support getting something from codec
+	// so response is stored to internal structure and dumped into trace messages
+	if (response_size) {
+		comp_dbg(dev, "waves_effect_message() trace response");
+		trace_array(dev, (const uint32_t *)waves_codec->response,
+			response_size / sizeof(uint32_t));
+	}
+
+	return 0;
+}
+
+static int waves_effect_config(struct comp_dev *dev, enum codec_cfg_type type)
+{
+	struct codec_config *cfg;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct codec_param *param;
+	uint32_t index;
+	uint32_t param_number = 0;
+	int ret = 0;
+
+	comp_info(dev, "waves_codec_configure() start type %d", type);
+
+	cfg = (type == CODEC_CFG_SETUP) ? &codec->s_cfg : &codec->r_cfg;
+
+	comp_info(dev, "waves_codec_configure() config %p, size %d, avail %d",
+		cfg->data, cfg->size, cfg->avail);
+
+	if (!cfg->avail || !cfg->size) {
+		comp_err(dev, "waves_codec_configure() no config for type %d, avail %d, size %d",
+			type, (unsigned int)cfg->avail, (unsigned int)cfg->size);
+		return -EIO;
+	}
+
+	if (cfg->size > MAX_CONFIG_SIZE_BYTES) {
+		comp_err(dev, "waves_codec_configure() size is too big %d", cfg->size);
+		return -EIO;
+	}
+
+	// incoming data in cfg->data is arranged according to struct codec_param
+	// there migh be more than one struct codec_param inside cfg->data, glued back to back
+	for (index = 0; index < cfg->size && (!ret); param_number++) {
+		uint32_t param_data_size;
+
+		param = (struct codec_param *)((char *)cfg->data + index);
+		param_data_size = param->size - sizeof(param->size) - sizeof(param->id);
+
+		comp_info(dev, "waves_codec_configure() param num %d id %d size %d",
+			param_number, param->id, param->size);
+
+		switch (param->id) {
+		case PARAM_NOP:
+			comp_info(dev, "waves_codec_configure() NOP");
+			ret = 0;
+			break;
+		case PARAM_MESSAGE:
+			ret = waves_effect_message(dev, param->data, param_data_size);
+			break;
+		case PARAM_REVISION:
+			ret = waves_effect_revision(dev);
+			break;
+		default:
+			ret = -EIO;
+			break;
+		}
+
+		index += param->size;
+	}
+
+	if (ret)
+		comp_err(dev, "waves_codec_configure() error %d", ret);
+	else
+		comp_dbg(dev, "waves_codec_configure() done");
+	return ret;
+}
+
+static int waves_effect_setup_config(struct comp_dev *dev)
+{
+	// apply setup config if present
+	struct codec_data *codec = comp_get_codec(dev);
+	int ret = 0;
+
+	comp_dbg(dev, "waves_effect_setup_config() start");
+
+	if (!codec->s_cfg.avail && !codec->s_cfg.size) {
+		comp_err(dev, "waves_effect_startup_config() no setup config");
+		return 0;
```

Comment:
if you are logging an error you should return an error

Reply:
fixed, thank you